experimental/grpc_attestation/src/attestation.rs

```diff
@@ -0,0 +1,210 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::{anyhow, Context};
+use futures::{Stream, StreamExt};
+use log::info;
+use oak_attestation_common::{
+    crypto::{AeadEncryptor, DiffieHellmanNegotiator},
+    get_sha256,
+    report::{AttestationInfo, Report},
+};
+use oak_grpc_attestation::proto::{
+    get_attested_data_request::RequestType, get_attested_data_response::ResponseType,
+    grpc_attestation_server::GrpcAttestation, AttestationResponse, GetAttestedDataRequest,
+    GetAttestedDataResponse,
+};
+use std::pin::Pin;
+use tonic::{Request, Response, Status, Streaming};
+
+/// Utility structure that receives messages from gRPC streams.
+struct Receiver {
+    request_stream: Streaming<GetAttestedDataRequest>,
+}
+
+impl Receiver {
+    /// Receives requests from [`Receiver::request_stream`].
+    /// `Ok(None)` output means that the corresponding gRPC stream has ended.
+    async fn receive(&mut self) -> anyhow::Result<Option<GetAttestedDataRequest>> {
+        let request = if let Some(request) = self.request_stream.next().await {
+            Some(request.context("Couldn't receive request")?)
+        } else {
+            None
+        };
+        Ok(request)
+    }
+}
+
+struct RequestHandler {
+    // Decrypts and encrypts messages using a Diffie-Hellman session key.
+    encryptor: AeadEncryptor,
+    // Processes data from client requests and creates responses.
+    handler: fn(&[u8]) -> Vec<u8>,
+}
+
+impl RequestHandler {
+    pub fn new(encryptor: AeadEncryptor, handler: fn(&[u8]) -> Vec<u8>) -> Self {
+        Self { encryptor, handler }
+    }
+
+    /// Decrypts a client request, runs [`RequestHandler::handler`] on them and returns an encrypted
+    /// response.
+    pub async fn handle_request(
+        &mut self,
+        receiver: &mut Receiver,
+    ) -> anyhow::Result<Option<GetAttestedDataResponse>> {
+        if let Some(request) = receiver
+            .receive()
+            .await
+            .context("Couldn't receive request")?
+        {
+            let request_type = request.request_type.context("Couldn't read request type")?;
+            let data = if let RequestType::Data(data) = request_type {
+                Ok(data)
+            } else {
+                Err(anyhow!("Received incorrect message type"))
+            }?;
+
+            let decrypted_data = self
+                .encryptor
+                .decrypt(&data)
```

Comment:
does the record format for data specify which direction the data is traveling in? for e.g. is it possible for the untrusted network to take data encrypted by the server and feed it back to itself which is then interpreted as data encrypted by the client?

Reply:
There are separate Protobuf messages for client requests and server responses, but their headers are neither encrypted not authenticated with AEAD. Only the data inside them is authenticated and encrypted.